examples/data-frames.dx

```diff
@@ -284,15 +235,62 @@ def crossJoin {n m f f'} (t:n=>{&...f}) (t':m=>{&...f'}) : (n & m)=>{...f & ...f
 -- It should be sufficient to do the join on a single column. One can always
 -- restructure the table so that the join happens on a tuple of values from
 -- the flattened columns.
-def leftJoin = todo
--- def leftJoin (c:Label) (l:n=>{@c: a & ...f}) (r:m=>{@c: a & ...f'})
---       : n=>{extra: Maybe {...f'} & ...f} = todo
+
+def findInUnordered {a n} [Eq a] (xs:n=>a) (v:a) : Maybe n =
+  -- Wasteful, but at least the parallelism is exposed.
+  (AsList num_found found_table) = argFilter (\x. x == v) xs
+  case num_found == 0 of
+    True -> Nothing
+    False -> Just found_table.(unsafeFromOrdinal _ 0)
+
+-- The implementation below takes (n * m) time, but could be done in O(n log m)
+-- if table r was indexed by a set of values of type a
+
+def leftJoin {n m a f f'} [Eq a]
+  (c:Label)
+  (left: n=>{@c: a & ...f})
+  (right:m=>{@c: a & ...f'})
+  : n=>{@c: a & extra: Maybe { &...f'} & ...f} =
+    right_c_vals = getColumn c right
+    for i:n.
+      {@c=left_c_val, ...rest_left} = left.i
+      newdata = case findInUnordered right_c_vals left_c_val of
+        Nothing -> Nothing
+        Just j ->
+          {@c=_, ...rest_right} = right.j
+          Just rest_right
+      {@c=left_c_val, extra=newdata, ...rest_left}
+
+:p leftJoin ##name students gradebook
+
+-- TODO: Match the example below exactly once multiple fields are supported.
+--:p leftJoin students gradebook {name:_ ? age:_}
+
+
 -- > leftJoin(students, gradebook, ["name", "age"])
 -- | name    | age | favorite color | quiz1 | quiz2 | midterm | quiz3 | quiz4 | final |
 -- | ------- | --- | -------------- | ----- | ----- | ------- | ----- | ----- | ----- |
 -- | "Bob"   | 12  | "blue"         | 8     | 9     | 77      | 7     | 9     | 87    |
 -- | "Alice" | 17  | "green"        | 6     | 8     | 88      | 8     | 7     | 85    |
 -- | "Eve"   | 13  | "red"          | 7     | 9     | 84      | 8     | 8     | 77    |
+
+-- The original example is ill-formed, because employees has deptID as Maybe Int,
+-- while departments has deptID as Int.  So we need to fix the tables up a bit.
```

Comment:
I don't think that it's ill formed in SQL where a join will just skip the missing values (since `== NULL` is always false), but in Dex you do have to fix up the types